master/buildbot/process/log.py

```diff
@@ -81,7 +83,12 @@ def addRawLines(self, lines):
         # formatted for the log type, and newline-terminated
         assert lines[-1] == '\n'
         assert not self.finished
-        yield self.lock.run(lambda: self.master.data.updates.appendLog(self.logid, lines))
+        if not self.truncated:
+            self.numChars = self.numChars + len(lines)
+            if self.master.config.logMaxSize is not None and self.numChars > self.master.config.logMaxSize:
```

Comment:
I think it makes sense to truncate at line granularity. Right now we truncate the granularity of a single batch submitted to `addRawLines`. If batches of different lengths were submitted, then we would truncate at different points even if the logs themselves were identical which would be really confusing to humans reading the logs.

To not to regress on performance it makes sense to only do anything special if self.master.config.logMaxSize is not None and split the incoming `lines` into a list of actual lines whenever we detect that the current chunk will be truncated somewhere.

Reply:
Thx for the suggestions, but I think the patch has a more severe problem: buildbot still aborts if large log files are sent, more info at https://github.com/buildbot/buildbot/issues/4504#issuecomment-451194269

If I can't fix that, it makes no sense to fix this regression here.